server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -588,6 +596,44 @@ private static void parseNonDynamicArray(DocumentParserContext context, final St
                 parseValue(context, lastFieldName);
             }
         }
+        postProcessDynamicMappers(context);
```

Comment:
Looking at this more. It seems we only have two options:
 - Do a top level post process, which ignores contexts and requires iterating over all fields. This doesn't fit with the current design (why I suggested moving towards things into the DocumentParser).
 - Add information to the `DocumentParserContext` that: keeps track when its within an array (marking on array entrance, exit, etc.), just like subobject parsers, and then change `Concrete#newDynamicDoubleField` to add things to a "Must collapse later" list. Then Post-process can drain that list, collapse, and add the correct mapping, or just forward them to add dynamic mapper via `addDynamicMapper`.

Reply:
Sorry for the back and forth. I think if `newDynamicDoubleField` knows its part of an array parse (as a value in that array), it could add itself to a "postProcessDynamicMappers" list which gets drained by `postProcessDynamicMappers(context);`, deleting each item in the list and either:

 - using that information to create a new dynamic mapper (if the count is correct)
 - pass the values forward to `dynamicMapper(value)` to handle it

Reply:
Hi @benwtrent I've spent some time looking at this today and it's still not clicking how this will solve the issue of parent-child fields. It seems to me that we may be able to make this more efficient by identifying only those fields that are part of an array before we do any post-processing, but I'm not quite connecting the dots for how this fixes sub-objects. Would you please be able to clarify a bit, or perhaps talk about it outside of this issue? Thank you!

Reply:
Per our discussion, I'll resolve this as we're going a different way with dynamic mappers.